cluster/src/main/scala/io/snappydata/gemxd/SparkSQLPrepareImpl.scala

```diff
@@ -0,0 +1,226 @@
+    def allParams(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+      case bl@BinaryComparison(left: Expression, ParamLiteral(Row(pos: Int), NullType, 0)) =>
+        addParamLiteral(pos, left.dataType, left.nullable, result)
+        bl
+      case blc@BinaryComparison(left: Expression,
+      Cast(ParamLiteral(Row(pos: Int), NullType, 0), _)) =>
+        addParamLiteral(pos, left.dataType, left.nullable, result)
+        blc
+      case ble@BinaryComparison(left: Expression, CaseWhen(branches, elseValue)) =>
+        handleCase(branches, elseValue, left.dataType, left.nullable, result)
+        ble
+      case blce@BinaryComparison(left: Expression, Cast(CaseWhen(branches, elseValue), _)) =>
+        handleCase(branches, elseValue, left.dataType, left.nullable, result)
+        blce
+      case br@BinaryComparison(ParamLiteral(Row(pos: Int), NullType, 0), right: Expression) =>
+        addParamLiteral(pos, right.dataType, right.nullable, result)
+        br
+      case brc@BinaryComparison(Cast(ParamLiteral(Row(pos: Int), NullType, 0), _),
+      right: Expression) =>
+        addParamLiteral(pos, right.dataType, right.nullable, result)
+        brc
+      case bre@BinaryComparison(CaseWhen(branches, elseValue), right: Expression) =>
+        handleCase(branches, elseValue, right.dataType, right.nullable, result)
+        bre
+      case brce@BinaryComparison(Cast(CaseWhen(branches, elseValue), _), right: Expression) =>
+        handleCase(branches, elseValue, right.dataType, right.nullable, result)
+        brce
+      case l@Like(left: Expression, ParamLiteral(Row(pos: Int), NullType, 0)) =>
+        addParamLiteral(pos, left.dataType, left.nullable, result)
+        l
+      case lc@Like(left: Expression, Cast(ParamLiteral(Row(pos: Int), NullType, 0), _)) =>
+        addParamLiteral(pos, left.dataType, left.nullable, result)
+        lc
+      case inlist@org.apache.spark.sql.catalyst.expressions.In(value: Expression,
+      list: Seq[Expression]) =>
+        list.map {
+          case ParamLiteral(Row(pos: Int), NullType, 0) =>
+            addParamLiteral(pos, value.dataType, value.nullable, result)
+          case Cast(ParamLiteral(Row(pos: Int), _, 0), _) =>
+            addParamLiteral(pos, value.dataType, value.nullable, result)
+          case x => x
+        }
+        inlist
+    }
+    handleSubQuery(allParams(plan), allParams)
+  }
+
+  def remainingParamLiterals(plan: LogicalPlan, result: mutable.HashSet[ParamLiteral]): Unit = {
+    def allParams(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
+      case c@Cast(ParamLiteral(Row(pos: Int), NullType, 0), castType: DataType) =>
+        addParamLiteral(pos, castType, false, result)
+        c
+      case cc@Cast(CaseWhen(branches, elseValue), castType: DataType) =>
+        handleCase(branches, elseValue, castType, false, result)
+        cc
+    }
+    handleSubQuery(allParams(plan), allParams)
+  }
+
+  def handleSubQuery(plan: LogicalPlan,
```

Comment:
Any level of nested subquery will automatically get handled with the pattern mentioned above.